org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java

```diff
@@ -570,288 +567,64 @@ public String[] getCommands (int cmdType)
-        			conn = m_ds.getConnection();
-        		}
-
-                if (conn != null)
-                {
-                    if (conn.getTransactionIsolation() != transactionIsolation)
-                        conn.setTransactionIsolation(transactionIsolation);
-                    if (conn.getAutoCommit() != autoCommit)
-                        conn.setAutoCommit(autoCommit);
-                }
-            }
-            catch (Exception e)
-            {
-                exception = e;
-                conn = null;
-                if (DBException.isInvalidUserPassError(e))
-                {
-                	//log might cause infinite loop since it will try to acquire database connection again
-                	/*
-                    log.severe("Cannot connect to database: "
-                        + getConnectionURL(connection)
-                        + " - UserID=" + connection.getDbUid());
-                    */
-                	StringBuilder msgerr = new StringBuilder("Cannot connect to database: ")
-                									.append(getConnectionURL(connection))
-                									.append(" - UserID=").append(connection.getDbUid());
-                	System.err.println(msgerr.toString());
-                }
-            }
-
-            if (conn == null && exception != null)
-            {
-            	//log might cause infinite loop since it will try to acquire database connection again
-            	/*
-                log.log(Level.SEVERE, exception.toString());
-                log.fine(toString()); */
-            	System.err.println(exception.toString());
-            }
-        }
-        catch (Exception e)
-        {
-            exception = e;
-        }
+		if (conn.getTransactionIsolation() != transactionIsolation)
+		{
+			conn.setTransactionIsolation(transactionIsolation);
+		}
+		if (conn.getAutoCommit() != autoCommit) 
+		{
+			conn.setAutoCommit(autoCommit);
+		}
 
-        try
-        {
-        	if (conn != null) {
-        		boolean trace = "true".equalsIgnoreCase(System.getProperty("org.adempiere.db.traceStatus"));
-        		int numConnections = m_ds.getNumBusyConnections();
-        		if (numConnections > 1)
-        		{
-	    			if (trace)
-	    			{
-	    				log.warning(getStatus());
-	    			}
-	    			if(numConnections >= m_maxbusyconnections && m_maxbusyconnections > 0)
-		            {
-	    				if (!trace)
-	    					log.warning(getStatus());
-		                //hengsin: make a best effort to reclaim leak connection
-		                Runtime.getRuntime().runFinalization();
-		            }
-        		}
-        	} else {
-        		//don't use log.severe here as it will try to access db again
-        		System.err.println("Failed to acquire new connection. Status=" + getStatus());
-        	}
-        }
-        catch (Exception ex)
-        {
-        }
-        if (exception != null)
-            throw exception;
-        return conn;
-    }   //  getCachedConnection
+		return conn;

```

Comment:
this is probably over simplification. we do want to log the exception and keep the "org.adempiere.db.traceStatus" feature.

Reply:
I've re-added the trace function.

The exception is already thrown so why also log it here? Throwing and logging is usually seen as an anti-pattern. It leads to multiple log entries for the same cause since the exception is also going to get logged somewhere higher up. And if higher up it is also "thrown-and-logged" the problem gets even bigger. It makes troubleshooting harder.

Most of the code which I removed was removed because didn't think it was needed anymore. For example the random sleep to wait for a free connection (`int randomNum = rand.nextInt(40 - 20 + 1) + 20;` etc.) is not needed with HikariCP because it will simply block until a connection is available. If you want threads to wait longer for a connection to free up then it's simply a matter of increasing the timeout. And for example the `Runtime.getRuntime().runFinalization();` part was removed because it only hints the GC to call the finalizers _next_ GC cycle. It doesn't free up anything immediately. So the benefit will be minimal.